File: C-Sharp/Data Structures and Algorithms/Trees/Trees/Trees/Tree.cs
using System;
using System.Collections.Generic;

namespace Trees
{
    public class Tree<T>
    {
        public Node<T> Root { get; set; }
        public Tree() { }
        public Tree(T value)
        {
            Node<T> node = new Node<T>(value);
            Root = node;
        }
        /// PreOrder
        #region
        /// <summary>
        /// Returns an array of the values contained in the tree ordered by the recursive pattern: root, right child, left child
        /// </summary>
        /// <param name="root">The node that will act as the upper root for the tree traversal</param>
        /// <returns>Ordered array of the tree's values starting from the given root</returns>
        public T[] PreOrder(Node<T> root)
        {
            List<T> list = new List<T>();
            PreOrder(list, root);
            return list.ToArray();
        }

        /// <summary>
        /// Indirect recursive method associated with its overload counterpart
        /// </summary>
        /// <param name="list">List of ordered values that is added to with each recursion</param>
        /// <param name="node">Current root node</param>
        private void PreOrder(List<T> list, Node<T> node)
        {
            list.Add(node.Value);
            if (node.LeftChild != null)
            {
                PreOrder(list, node.LeftChild);
            }
            if (node.RightChild != null)
            {
                PreOrder(list, node.RightChild);
            }
        }
        #endregion

        /// InOrder
        #region
        /// <summary>
        /// Returns an array of the values contained in the tree ordered by the recursive pattern: right child, root, left child
        /// </summary>
        /// <param name="root">The node that will act as the upper root for the tree traversal</param>
        /// <returns>Ordered array of the tree's values starting from the given root</returns>
        public T[] InOrder(Node<T> root)
        {
            List<T> list = new List<T>();
            InOrder(list, root);
            return list.ToArray();
        }

        /// <summary>
        /// Indirect recursive method associated with its overload counterpart
        /// </summary>
        /// <param name="list">List of ordered values that is added to with each recursion</param>
        /// <param name="node">Current root node</param>
        private void InOrder(List<T> list, Node<T> node)
        {
            if (node.LeftChild != null)
            {
                InOrder(list, node.LeftChild);
            }
            list.Add(node.Value);
            if (node.RightChild != null)
            {
                InOrder(list, node.RightChild);
            }
        }
        #endregion

        /// PostOrder
        #region
        /// <summary>
        /// Returns an array of the values contained in the tree ordered by the recursive pattern: right child, left child, root
        /// </summary>
        /// <param name="root">The node that will act as the upper root for the tree traversal</param>
        /// <returns>Ordered array of the tree's values starting from the given root</returns>
        public T[] PostOrder(Node<T> root)
        {
            List<T> list = new List<T>();
            PostOrder(list, root);
            return list.ToArray();
        }

        /// <summary>
        /// Indirect recursive method associated with its overload counterpart
        /// </summary>
        /// <param name="list">List of ordered values that is added to with each recursion</param>
        /// <param name="node">Current root node</param>
        private void PostOrder(List<T> list, Node<T> node)
        {
            if (node.LeftChild != null)
            {
                PostOrder(list, node.LeftChild);
            }
            if (node.RightChild != null)
            {
                PostOrder(list, node.RightChild);
            }
            list.Add(node.Value);
        }
        #endregion
    }
}

File: Java/Data Structures and Algorithms/dsas/Trees/src/main/java/tree/Tree.java
package tree;

import java.util.ArrayList;
import java.util.List;

public class Tree {
    public Node root;
    public Tree() { }
    public Tree(int value) {
        Node node = new Node(value);
        root = node;
    }

    public Node getRoot() { return root; }
    public void setRoot(Node root) { this.root = root; }

    public Object[] PreOrder(Node root) {
        List<Integer> list = new ArrayList<>();
        PreOrder(list, root);
        return list.toArray();
    }

    public Object[] PreOrder() {
        Node root = this.root;
        List<Integer> list = new ArrayList<>();
        PreOrder(list, root);
        return list.toArray();
    }

    private void PreOrder(List<Integer> list, Node node) {
        list.add(node.value);
        if (node.leftChild != null)
            PreOrder(list, node.leftChild);
        if (node.rightChild != null)
            PreOrder(list, node.rightChild);
    }

    public Object[] InOrder(Node root) {
        List<Integer> list = new ArrayList<>();
        InOrder(list, root);
        return list.toArray();
    }

    public Object[] InOrder() {
        Node root = this.root;
        List<Integer> list = new ArrayList<>();
        InOrder(list, root);
        return list.toArray();
    }

    private void InOrder(List<Integer> list, Node node) {
        if (node.leftChild != null)
            InOrder(list, node.leftChild);
        list.add(node.value);
        if (node.rightChild != null)
            InOrder(list, node.rightChild);
    }

    public Object[] PostOrder(Node root) {
        List<Integer> list = new ArrayList<>();
        PostOrder(list, root);
        return list.toArray();
    }

    public Object[] PostOrder() {
        Node root = this.root;
        List<Integer> list = new ArrayList<>();
        PostOrder(list, root);
        return list.toArray();
    }

    private void PostOrder(List<Integer> list, Node node) {
        if (node.leftChild != null)
            PostOrder(list, node.leftChild);
        if (node.rightChild != null)
            PostOrder(list, node.rightChild);
        list.add(node.value);
    }

    public int findMax() {
        if (root == null) throw new NullPointerException();
        return findMax(this.root, this.root.value);
    }

    private int findMax(Node node, int max) {
        if (node.value > max) {
            max = node.value;
        if (node.leftChild != null)
            max = findMax(node.leftChild, max);
        if (node.rightChild != null)
            max = findMax(node.rightChild, max);
        }
        return max;
    }

}
